core/tee/tee_enc_fs.c

```diff
@@ -0,0 +1,731 @@
+	}
+
+	if (!fdp || !buf)
+		goto exit;
+
+	if (fdp->flags & TEE_FS_O_WRONLY)
+		goto exit;
+
+	remain = fdp->len - fdp->pos;
+	if (len > remain)
+		len = remain;
+
+	memcpy(buf, fdp->data + fdp->pos, len);
+	res = len;
+
+exit:
+	return res;
+}
+
+static int tee_fs_write(int fd, const void *buf, size_t len)
+{
+	int res = -1;
+	size_t remain;
+	struct tee_fs_fd *fdp = handle_lookup(&fs_handle_db, fd);
+
+	DMSG("write, fd=%d, buf=%p, len=%d", fd, buf, len);
+
+	if (len == 0) {
+		res = 0;
+		goto exit;
+	}
+
+	if (!buf)
+		goto exit;
+
+	if (!fdp)
+		goto exit;
+
+	/* restrict this to avoid file hole */
+	if (fdp->pos > fdp->len)
+		goto exit;
+
+	if (fdp->flags & TEE_FS_O_RDONLY)
+		goto exit;
+
+	/* check if we need to update file size */
+	remain = fdp->len - fdp->pos;
+	if (len > remain) {
+		size_t new_file_len = fdp->len + (len - remain);
+
+		res = update_file_size(fdp, new_file_len);
+		if (res != 0)
+			goto exit;
+	}
+
+	memcpy(fdp->data + fdp->pos, buf, len);
+	fdp->pos += len;
+	res = len;
+
+exit:
+	return res;
+}
+
+static tee_fs_off_t tee_fs_lseek(int fd, tee_fs_off_t offset, int whence)
+{
+	tee_fs_off_t res = -1;
+	tee_fs_off_t new_pos;
+	struct tee_fs_fd *fdp = handle_lookup(&fs_handle_db, fd);
+
+	if (!fdp)
+		goto exit;
+
+	switch (whence) {
+
+	case TEE_FS_SEEK_SET:
+		new_pos = offset;
+		break;
+
+	case TEE_FS_SEEK_CUR:
+		new_pos = fdp->pos + offset;
+		break;
+
+	case TEE_FS_SEEK_END:
+		new_pos = fdp->len + offset;
+		break;
+
+	default:
+		goto exit;
+	}
+
+	/*
+	 * file hole is not supported in this implementation,
+	 * restrict the file postion within file length
+	 * for simplicity
+	 */
+	if (new_pos > (tee_fs_off_t)fdp->len)
+		goto exit;
+
+	res = fdp->pos = new_pos;
+exit:
+	return res;
+}
+
+static int tee_fs_rename(const char *old, const char *new)
+{
+	int res = -1;
+	char *tmp = NULL;
+	struct tee_fs_rpc head = { 0 };
+	size_t len_old = strlen(old) + 1;
+	size_t len_new = strlen(new) + 1;
+	size_t len = len_old + len_new;
+
+	if (len > TEE_FS_NAME_MAX)
+		goto exit;
+
+	tmp = malloc(len);
+	if (!tmp)
+		goto exit;
+	memcpy(tmp, old, len_old);
+	memcpy(tmp + len_old, new, len_new);
+
+	head.op = TEE_FS_RENAME;
+
+	res = tee_fs_send_cmd(&head, tmp, len, TEE_FS_MODE_IN);
+	if (!res)
+		res = head.res;
+
+exit:
+	free(tmp);
+	return res;
+}
+
+static int tee_fs_unlink(const char *file)
+{
+	int res = -1;
+	struct tee_fs_rpc head = { 0 };
+	size_t len = strlen(file) + 1;
+
+	if (len > TEE_FS_NAME_MAX)
+		goto exit;
+
+	head.op = TEE_FS_UNLINK;
+
+	res = tee_fs_send_cmd(&head, (void *)file, len, TEE_FS_MODE_IN);
+	if (!res)
+		res = head.res;
+
+exit:
+	return res;
+}
+
+static int tee_fs_ftruncate(int fd, tee_fs_off_t length)
+{
+	int res = -1;
+	struct tee_fs_fd *fdp = handle_lookup(&fs_handle_db, fd);
+
+	if (!fdp)
+		goto exit;
+
+	if (fdp->flags & TEE_FS_O_RDONLY)
+		goto exit;
+
+	res = update_file_size(fdp, length);
+
+exit:
+	return res;
+}
+
+static int tee_fs_mkdir(const char *path, tee_fs_mode_t mode)
```

Comment:
Many functions are the same, whether it is fs-only or enc_fs (`tee_fs_mkdir`,...). I think it is worth to commonalize as it will lead to cleaner code, as the `if (res) goto exit;` now fixed in enc_fs, but still not fixed in fs-only.